scipy/optimize/_zeros_py.py

```diff
@@ -1388,3 +1388,278 @@ def toms748(f, a, b, args=(), k=1,
+        lower_fun_val : float
+            The lower function value.
+        upper_fun_val : float
+            The upper function value.
+
+    Notes
+    -----
+    As written in [Sachs2015]_ which in turn is based on Chandrupatla's
+    algorithm as described in [Scherer2018]_.
+
+    References
+    ----------
+
+    .. [Sachs2015]
+       Sachs, Jason
+       "Ten Little Algorithms, Part 5: Quadratic Extremum Interpolation and
+       Chandrupatla's Method",
+       https://www.embeddedrelated.com/showarticle/855.php
+
+    .. [Scherer2018]
+       Scherer, Philipp
+       *Computational physics: Simulation of classical and Quantum Systems*.
+       SPRINGER INTERNATIONAL PU, 2018. Ch. 6.1.7.3.
+
+    .. [Chandrupatla1997]
+        Chandrupatla, Tirupathi R.
+        "A new hybrid quadratic/bisection algorithm for finding the zero of a
+        nonlinear function without using derivatives".
+        Advances in Engineering Software, 28(3), 145-149.
+        https://doi.org/10.1016/s0965-9978(96)00051-8
+
+    See Also
+    --------
+    brentq, brenth, ridder, bisect, newton
+
+    Examples
+    --------
+    >>> from scipy import optimize
+    >>> def f(x,a):
+    ...     return a-x*x
+
+    >>> y = optimize._chandrupatla(f,0,3,args=(7.0,))
+    >>> y.x
+    2.6457513110645907
+
+    >>> x0 = np.zeros((2,1,1))
+    >>> x1 = np.full((1,2,1), 3)
+    >>> k = np.array([[[1, 2]]])
+    >>> res = optimize._chandrupatla(f,x0=x0,x1=x1,
+                              args=(k,),
+                              maxiter=50)
+    >>> res.x
+    [[[ 1.000e+00  1.414e+00]
+         [ 1.000e+00  1.414e+00]]
+
+        [[ 1.000e+00  1.414e+00]
+         [ 1.000e+00  1.414e+00]]]
+
+    permission from Jason Sachs (@jason-s):
+        https://github.com/scipy/scipy/issues/7242#issuecomment-1314178133
+    """
+
+    # adapted from an earlier implementation of Jason Sachs'
+    # as written in https://www.embeddedrelated.com/showarticle/855.php
+    # which in turn is based on Chandrupatla's algorithm as described in
+    # Scherer https://books.google.com/books?id=cC-8BAAAQBAJ&pg=PA95
+
+    # Initialization
+    b = x0
+    a = c = x1
+    fb = f(b, *args)
+    fa = fc = f(a, *args)
+    t = 0.5
+    iterations = 0
+
+    a, b, fa, fb = np.broadcast_arrays(a, b, fa, fb)
+    intermediate_shape = a.shape
+    a, b, fa, fb = np.atleast_1d(a, b, fa, fb)
+    shape = a.shape
+
+    iterations = np.zeros_like(a)
+
+    # flag to check state of convergence
+    flag = np.full(shape, _EINPROGRESS)
+
+    # Initialize an array of False,
+    # determines whether we should do inverse quadratic interpolation
+    iqi = np.zeros(shape, dtype=bool)
+
+    funcalls = np.full(np.shape(flag), 0)
+    terminate = False
+
+    while iterations.any() < maxiter:
+        # use t to linearly interpolate between a and b,
+        # and evaluate this function as our newest estimate xt
+        xt = a + t * (b - a)
+        ft = f(xt, *args)
```

Comment:
Can you add an argument for testing purposes, maybe "`ensure_scalar_x`", that allows us to ensure that `f` is called with a scalar instead of an array?

For instance, maybe at the top:
```python3
if ensure_scalar_x:
    def f_wrapped(x, *args, **kwargs):
         return np.asarray([f(x[0], *args, **kwargs)])
else:
    f_wrapped = f
```
then you use `f_wrapped` throughout instead of `f`. This allows you to continue to use `x` as an array throughout; you only need to branch here when re-defining the function instead of in the main logic of the code.